### ISSUE TITLE
ASDF extensions for custom compression types?

### ISSUE THREAD

**Issue by lgarrison** [asdf-format/asdf#926]
I have ASDF files where the binary data is compressed with a custom compression, identified by the `'blsc'` tag in the binary header.  The current way I'm recommending my users read these files is by downloading my fork of ASDF that knows how to do this decompression.  But this is definitely a fragile solution; if a user accidentally installs mainline ASDF, perhaps because my fork doesn't satisfy some ASDF version requirement for another project, then they can't read my files and have to go find my fork again.

I started reading about the ASDF extension mechanism, and I was wondering if that might be a path to a more robust solution.  In particular, I see that extension classes can be registered with setuptools entry points to inform ASDF how to process custom data types, and I was wondering if custom compression types might be registered in the same way.  Does that make sense in the context of ASDF extensions, or would that be orthogonal to their purpose?

Thanks!

**Comment by eslavich** [comment]
I think it's a great idea to have a entry points plugin for additional compression types.  That could be added to the existing Extension class or a new entry point dedicated to compression providers.

**Comment by lgarrison** [comment]
Thanks; I'd be happy to work on a PR for this.  My particular use-case is a little strange, though, because the files are already written without any reference to such an extension, so there's no `history: extensions: - lgarrison_blosc` or whatever in the YAML portion of the files.  The files are too numerous to rewrite, unfortunately.  Perhaps a way forward would be to provide a runtime flag to opt-in to "eager" or "unsafe" use of compression extensions that just rely on the compression name.  For example, in my Python package through which users will read my files, I could write:
```python
import asdf
asdf.compression.allow_unchecked_extensions()
```

Then, the future file loads will be checked against the list of compression types registered via the entry points. Any matching names will be decompressed via that extension.  If `allow_unchecked_extensions()` is not used, then any decompression must be done by an extension that is explicitly listed in the YAML of that file (which I think is how the type extension mechanism works, if I understand correctly).

**Comment by eslavich** [comment]
Maybe it would be generally safe to rely on the compression name if there aren't any conflicts among the installed extensions?  If not, we do have a new [configuration feature](https://github.com/asdf-format/asdf/blob/master/asdf/config.py#L208) that would be an appropriate place to add an `allow_unchecked_extensions` flag.

I was thinking about this over the weekend and I'm in favor of adding a list of compressors to the [new Extension class](https://github.com/asdf-format/asdf/blob/master/asdf/extension/_extension.py#L11).  We would certainly welcome a PR!  @perrygreenfield is this reasonable as a stepping stone towards the ideas discussed in https://github.com/asdf-format/asdf-standard/issues/248?

**Comment by perrygreenfield** [comment]
I think so, but I wonder if there is some way of using a name convention that indicates that it is custom or provision in some way. That would be a pain for the pioneer in that they now have a bunch of files with compression label that may be promoted to a standard one (perhaps some alias mechanism can be permitted in the future adoption, i.e., not only do I recognize compression_xxx, I see acme.edu/compression_xxx as the same thing).

**Comment by lgarrison** [comment]
I'm looking at the `Extension` source and trying to sketch out a plan for the PR.  It looks like the `Converter` mechanism *almost* works for compression, but has some concepts that seem tied to type serialization.  If I were to implement a `Compressor` mechanism in parallel to `Converter`, I think it might start with this:

- Add a `compressors()` property to `Extension` and `ExtensionProxy` classes in analogy with `converters()`
- Make `asdf/extensions/_compressor.py` with `Compressor` and `CompressorProxy` classes
- Instead of `tags`, `types`, `to_yaml_tree`, and `from_yaml_tree` properties, these classes will have `labels`, `compression_types`, `compress`, and `decompress`.

It's tempting to try to modify `Converter` instead or add a layer of abstraction so the common parts can be shared.  But that might abstract-ify things so much that it makes that whole part of the code harder to understand.  What do you think?

I haven't wrapped my head around the Proxy mechanism yet, is that something that's needed for compression as well as type serialization?

**Comment by eslavich** [comment]
> It's tempting to try to modify Converter instead or add a layer of abstraction so the common parts can be shared. But that might abstract-ify things so much that it makes that whole part of the code harder to understand. What do you think?

I think a new `Compressor` ABC is the way to go, `Converter` is intended for converting to and from YAML.

> I haven't wrapped my head around the Proxy mechanism yet, is that something that's needed for compression as well as type serialization?

I added the `ConverterProxy` and `ExtensionProxy` classes because I wanted to normalize the values returned by user extensions and give users a little more flexibility without having to deal with that flexibility everywhere in the code.  I bet you won't need a `CompressorProxy` since the `Compressor` properties are all straightforward and required.

> Instead of tags, types, to_yaml_tree, and from_yaml_tree properties, these classes will have labels, compression_types, compress, and decompress.

That sounds reasonable to me.  What's the distinction between `labels` and `compression_types`?

I think eventually we'll want to fetch the compressors from the `ExtensionManager` class, but that is hard to get ahold of from compression.py right now due to a deprecated feature that we have to support until asdf 3.0.  In the meantime I'd recommend grabbing all the enabled extensions from the config object (`asdf.config.get_config().extensions`).

**Comment by lgarrison** [comment]
I was thinking `labels` would be strings like `'blsc'`, and `compression_types` would be Python types like `lgarrison_pkg.asdf.BloscCompressor`.  I think that's redundant though... I guess all you need is to be able to identify what labels belong to what Compressor, and then functions like `compress()` and `decompress()` that know how to do the work.

Good to know about the `ExtensionManager`, I had been looking at that.  I'll grab the global config instead!

**Comment by lgarrison** [comment]
I'm writing a test for the compression extension feature (will open a PR soon), and I've come across a strange problem.  If I put a test in `test_entry_points.py` so that `monkeypatch_entry_points()` can be used to simulate the registration of a real compression extension, the test fails because it tries to access the internet for schema validation (trace below).  Adding this minimal test to `test_entry_points.py` is enough to trigger the issue:
```
def test_dummy():
    tree = dict(arr=[1,2,3])
    af = asdf.AsdfFile(tree=tree)
```

Any ideas?  Marking the test as `remote_data` just results in an HTTP 301 error, and also probably isn't the "right" solution anyway.


<details>

```
(forge) lgarrison@worker3113:~/asdf/asdf/tests$ pytest --ignore-glob='.ipynb_checkpoints' -x test_entry_points.py::test_dummy
============================================================================================================= test session starts =============================================================================================================
platform linux -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /mnt/home/lgarrison/asdf, inifile: setup.cfg
plugins: astropy-header-0.1.2, filter-subpackage-0.1.1, openfiles-0.5.0, remotedata-0.3.2, arraydiff-0.3, doctestplus-0.7.0, cov-2.8.1, hypothesis-5.15.0, asdf-2.8.0.dev159+gf8c1e18.d20210220
collected 1 item                                                                                                                                                                                                                              

test_entry_points.py F                                                                                                                                                                                                                  [100%]

================================================================================================================== FAILURES ===================================================================================================================
_________________________________________________________________________________________________________________ test_dummy __________________________________________________________________________________________________________________

    def test_dummy():
        tree = dict(arr=[1,2,3])
>       af = asdf.AsdfFile(tree=tree)

test_entry_points.py:152: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../asdf.py:174: in __init__
    self.tree = tree
../asdf.py:587: in tree
    self._validate(asdf_object, custom=bool(tree))
../asdf.py:617: in _validate
    schema.validate(tagged_tree, self, reading=reading)
../schema.py:658: in validate
    validator = get_validator(schema, ctx, validators, ctx.resolver,
../schema.py:563: in get_validator
    cls = _create_validator(validators=validators, visit_repeat_nodes=_visit_repeat_nodes)
../schema.py:263: in _create_validator
    meta_schema = _load_schema_cached(YAML_SCHEMA_METASCHEMA_ID, extension.get_default_resolver(), False, False)
../schema.py:464: in _load_schema_cached
    schema, url = loader(url)
../schema.py:375: in load_schema
    return _load_schema(url)
../schema.py:348: in _load_schema
    with generic_io.get_file(url) as fd:
../generic_io.py:1174: in get_file
    return _make_http_connection(init, mode, uri=uri)
../generic_io.py:1049: in _make_http_connection
    connection.connect()
../../../miniconda3/envs/forge/lib/python3.8/http/client.py:915: in connect
    self.sock = self._create_connection(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (('stsci.edu', 80), <object object at 0x15540e3d9760>, None), kwargs = {}, host = 'stsci.edu', addr_arg = 0, valid_hosts = {'127.0.0.1', 'localhost'}, hostname = 'worker3113', fqdn = 'worker3113.cm.cluster'
host_ips = {'130.167.220.64', '130.167.220.65', '130.167.220.66', '130.167.220.67', 'stsci.edu'}

    def new_function(*args, **kwargs):
        if isinstance(args[0], socket.socket):
            if not args[0].family in (socket.AF_INET, socket.AF_INET6):
                # Should be fine in all but some very obscure cases
                # More to the point, we don't want to affect AF_UNIX
                # sockets.
                return original_function(*args, **kwargs)
            host = args[1][0]
            addr_arg = 1
            valid_hosts = set(['localhost', '127.0.0.1', '::1'])
        else:
            # The only other function this is used to wrap currently is
            # socket.create_connection, which should be passed a 2-tuple, but
            # we'll check just in case
            if not (isinstance(args[0], tuple) and len(args[0]) == 2):
                return original_function(*args, **kwargs)
    
            host = args[0][0]
            addr_arg = 0
            valid_hosts = set(['localhost', '127.0.0.1'])
    
        # Astropy + GitHub data
        if allow_astropy_data:
            for valid_host in ASTROPY_HOSTS:
                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
    
        # Only GitHub data
        elif allow_github_data:
            for valid_host in GITHUB_HOSTS:
                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
    
        hostname = socket.gethostname()
        fqdn = socket.getfqdn()
    
        if host in (hostname, fqdn):
            host = 'localhost'
            host_ips = set([host])
            new_addr = (host, args[addr_arg][1])
            args = args[:addr_arg] + (new_addr,) + args[addr_arg + 1:]
        else:
            host_ips = _resolve_host_ips(host)
    
        if len(host_ips & valid_hosts) > 0:  # Any overlap is acceptable
            return original_function(*args, **kwargs)
        else:
>           raise IOError("An attempt was made to connect to the internet "
                          "by a test that was not marked `remote_data`. The "
                          "requested host was: {0}".format(host))
E           OSError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: stsci.edu

../../../miniconda3/envs/forge/lib/python3.8/site-packages/pytest_remotedata/disable_internet.py:99: OSError
=========================================================================================================== short test summary info ===========================================================================================================
FAILED test_entry_points.py::test_dummy - OSError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: stsci.edu
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================== 1 failed in 0.19s ==============================================================================================================
```
</details>

**Comment by eslavich** [comment]
The `monkeypatch_entry_points` fixture prevents the built-in extension from being registered, so the library is missing the core schemas.  I think it's fair to trust that the entry point infrastructure works as advertised since it's already well tested.  Instead of monkey patching the entry points method, I'd recommend something like this:

```python
from asdf import config_context

def test_extension_with_compression:
    with config_context() as config:
        class ExtensionWithCompression:
            # ...
        config.add_extension(ExtensionWithCompression())
        
        # test writing and reading a file with custom compression
```

The `config_context()` method creates a config environment that gets reset when the context is exited, and the `add_extension` method registers the custom extension in that environment so that it becomes available to your code in compression.py.

**Comment by eslavich** [comment]
Something I just realized is that we'll need some special handling to list extensions used for compression in the file's metadata.  It's a little tricky because currently that list is compiled before the tree is written, which happens before block writes, so we can't collect the information as we compress.  I haven't thought this through but maybe we can just run through the blocks here:

https://github.com/asdf-format/asdf/blob/master/asdf/asdf.py#L1001

and call `serialization_context._mark_extension_used(...)` on any extension whose compression code is set as the `output_compression` of a block.

**Comment by lgarrison** [comment]
Thanks, I think I got something along those lines working.  Is there a way we can hint to the user that a certain extension is needed to open a compression type?  Does the serialization context need to record the labels an extension can handle?

**Comment by eslavich** [comment]
>  Is there a way we can hint to the user that a certain extension is needed to open a compression type?

There is a warning issued by the library when a user opens a file that was written with an extension that isn't installed on their system.  The warning text includes the Python package name so that's a pretty good hint.  Someday we'd like to run a website with a searchable registry of extensions but it'll be a while before we get that going :)

> Does the serialization context need to record the labels an extension can handle?

The serialization context object keeps a reference to the extension itself, so it should have access to everything via that?

**Comment by lgarrison** [comment]
I remember seeing warnings about missing extensions in the past, but I'm not seeing it in this case.  The extension is in the history, but I get a compression validation failure and no warning.  Perhaps the validation is happening before ASDF has a chance to issue the warning?  In any case, I think this will be easier to debug with a concrete example, so I'll go ahead and get the PR going.

**Comment by eslavich** [comment]
Resolved by #931
